File: aloha-java/agent/src/main/java/com/aloha/a2a/agent/Tools.java
package com.aloha.a2a.agent;

import dev.langchain4j.agent.tool.Tool;
import jakarta.enterprise.context.ApplicationScoped;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.List;
import java.util.Random;
import java.util.stream.Collectors;

/**
 * Tools available to the Dice Agent.
 */
@ApplicationScoped
public class Tools {
    
    private static final Logger logger = LoggerFactory.getLogger(Tools.class);
    private final Random random = new Random();

    /**
     * Rolls an N-sided dice and returns the result.
     * 
     * @param sides The number of sides on the dice (must be positive)
     * @return The result of the dice roll (1 to N)
     */
    @Tool("Rolls an N-sided dice and returns a random number between 1 and N")
    public int rollDice(int sides) {
        if (sides <= 0) {
            logger.error("Invalid dice sides: {}", sides);
            throw new IllegalArgumentException("Dice must have at least 1 side");
        }
        
        int result = random.nextInt(sides) + 1;
        logger.info("Rolled {}-sided dice: {}", sides, result);
        return result;
    }

    /**
     * Checks which numbers in the list are prime.
     * 
     * @param numbers List of integers to check
     * @return A string describing which numbers are prime
     */
    @Tool("Checks if the given numbers are prime and returns which ones are prime")
    public String checkPrime(List<Integer> numbers) {
        if (numbers == null || numbers.isEmpty()) {
            return "No numbers provided to check.";
        }
        
        List<Integer> primes = numbers.stream()
                .filter(this::isPrime)
                .collect(Collectors.toList());
        
        if (primes.isEmpty()) {
            logger.info("No prime numbers found in: {}", numbers);
            return "None of the numbers are prime.";
        }
        
        String result = primes.stream()
                .map(String::valueOf)
                .collect(Collectors.joining(", ")) + " are prime numbers.";
        
        logger.info("Prime check for {}: {}", numbers, result);
        return result;
    }

    /**
     * Checks if a number is prime.
     * 
     * @param n The number to check
     * @return true if the number is prime, false otherwise
     */
    private boolean isPrime(int n) {
        if (n <= 1) {
            return false;
        }
        if (n == 2) {
            return true;
        }
        if (n % 2 == 0) {
            return false;
        }
        
        int sqrt = (int) Math.sqrt(n);
        for (int i = 3; i <= sqrt; i += 2) {
            if (n % i == 0) {
                return false;
            }
        }
        
        return true;
    }
}

File: aloha-python/server/tools.py
"""Tools available to the Dice Agent."""

import logging
import random
from typing import List

logger = logging.getLogger(__name__)


def roll_dice(sides: int) -> int:
    """
    Rolls an N-sided dice and returns the result.
    
    Args:
        sides: The number of sides on the dice (must be positive)
        
    Returns:
        The result of the dice roll (1 to N)
        
    Raises:
        ValueError: If sides is not positive
    """
    if sides <= 0:
        logger.error(f"Invalid dice sides: {sides}")
        raise ValueError("Dice must have at least 1 side")
    
    result = random.randint(1, sides)
    logger.info(f"Rolled {sides}-sided dice: {result}")
    return result


def check_prime(numbers: List[int]) -> str:
    """
    Checks which numbers in the list are prime.
    
    Args:
        numbers: List of integers to check
        
    Returns:
        A string describing which numbers are prime
    """
    if not numbers:
        return "No numbers provided to check."
    
    primes = [n for n in numbers if is_prime(n)]
    
    if not primes:
        logger.info(f"No prime numbers found in: {numbers}")
        return "None of the numbers are prime."
    
    result = ", ".join(str(p) for p in primes) + " are prime numbers."
    logger.info(f"Prime check for {numbers}: {result}")
    return result


def is_prime(n: int) -> bool:
    """
    Checks if a number is prime.
    
    Args:
        n: The number to check
        
    Returns:
        True if the number is prime, False otherwise
    """
    if n <= 1:
        return False
    if n == 2:
        return True
    if n % 2 == 0:
        return False
    
    sqrt_n = int(n ** 0.5)
    for i in range(3, sqrt_n + 1, 2):
        if n % i == 0:
            return False
    
    return True
